### DISCUSSION TITLE
Typescript event types for invoke.done

### DISCUSSION THREAD

**Discussion by djake** [statelyai/xstate#1645] [Q&A] [answered]
Long time lurker, first time user here. First, thank you so much for this library and all the associated work you do promoting it and explaining it.

I am using typescript. I have followed a suggestion at https://github.com/davidkpiano/xstate/discussions/1591 to use TS type assertions to ensure correctly typed actions.

I have added a service to do a fetch and I want to assign the results to context. Applying the suggestion above, it seems I need something like `{ type:' done.invoke.foo', data: FooData }` in my TEvents union type. I've got my code working by adding `DoneInvokeEvent<FooData>` and added `assertEventType(event, 'done.invoke.foo');` assertion inside the associated action. Everything appears to work, however, since I'm just getting going, I'm wondering if anyone can comment on whether this is a good approach, or if there is another best practice. In particular, I can't tell if my writing `done.invoke` is inappropriately making use of library internals.

P.S. As you may know, typescript 4.1 now includes [template literal types](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#template-literal-types), which seem like they might be useful for typing the invoke done types.

**Comment by djake** [comment]
My full POC code, in case this helps clarify:

```typescript
import { inspect } from '@xstate/inspect';
import { Machine, interpret, assign, EventObject, DoneInvokeEvent } from 'xstate';

inspect()

interface ConformanceStatement {}

type AppEvent =
| { type: 'CONFIGURE', serverHost: string, basePath: string }
| { type: 'CONFORMANCE', data: {} }
| DoneInvokeEvent<ConformanceStatement>

// The hierarchical (recursive) schema for the states
interface AppStateSchema {
    states: {
      initial: {};
      configured: {}
      ready: {};
    };
  }

// The context (extended state) of the machine
interface AppContext {
    serverHost: string;
    basePath: string;
    conformance: {}
}

// Actions
const configure = assign<AppContext, AppEvent>({
    serverHost: (context, event) => { 
        assertEventType(event, 'CONFIGURE');
        return event.serverHost;
    },
    basePath: (context, event) => {
        assertEventType(event, 'CONFIGURE');
        return event.basePath;
    },
  });

const conformance = assign<AppContext, AppEvent>({
    conformance: (context, event) => {
        assertEventType(event, 'done.invoke.getConformanceStatement');
        return event.data;
    }
})

//Services
function getConformanceStatement(context: AppContext) {
    return fetch(`${context.serverHost}${context.basePath}/metadata`).then(r => r.json())
}

const serverMachine = Machine<AppContext, AppStateSchema, AppEvent>({
    key: 'server',
    initial: 'initial',
    context: { 
        conformance: {},
        serverHost: '',
        basePath: ''
    },
    states: {
        initial: {
            on: {
                CONFIGURE: {
                    target: 'configured',
                    actions: 'configure'
                }
            }
        },
        configured: {
            invoke: {
                id: 'fetchConformanceStatement',
                src: 'getConformanceStatement',
                onDone: {
                    target: 'ready',
                    actions: 'conformance'
                },
                onError: {
                    target: 'configured'
                }
            },
        }, 
        ready: {
            type: 'final'
        }
    }
}, {
    actions: {
        configure,
        conformance
    },
    services: {
        getConformanceStatement,
    }
});

// test
const serverService = interpret(serverMachine, { devTools: true })

serverService.start()

/**
 * 
 * @see https://github.com/davidkpiano/xstate/discussions/1591
 */
function assertEventType<TE extends EventObject, TType extends TE["type"]>(
    event: TE,
    eventType: TType
  ): asserts event is TE & { type: TType } {
    if (event.type !== eventType) {
      throw new Error(
        `Invalid event: expected "${eventType}", got "${event.type}"`
      );
    }
  }
```

**Reply by Abekonge** [reply]
Thanks!

**Comment by UberMouse** [comment, marked as answer]
`DoneInvokeEvent` is not currently ideal for this. It doesn't specify a `type` for the event, it's just a `string` so it kind of defeats the type safety of the event union. With TS 4.1+ it would be easy to write one that also took the `id` of what was being invoked (ie `DoneInvokeEvent<"foo", PayloadType>` => `{type: "done.invoke.foo", data: PayloadType}`) but for now the way I handle this is just manually adding the event type into the union

ie `{type: "done.invoke.<id>", data: PayloadType}` as that is the underlying type of a DoneEvent.

**Reply by davidkpiano** [reply]
I could do that, but that would be a type-breaking change. How about `DoneInvokeEvent<PayloadType, Id>`?

EDIT: Perhaps this can be done in userland for now. We're waiting a couple of minor versions of TS before introducing new language features to the types.

**Reply by UberMouse** [reply]
Ya, that would work fine. Since it was an example I just did it without thinking about breaking changes and `Id` first makes the most sense.
